# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1762885321 -3600
# Parent  43329d8142e2025a4aa1a0fc8e443acc1929baaa
Feature: Allow search of all folders of all mail accounts. Feed and news accounts are excluded.

diff --git a/mail/base/content/SearchDialog.js b/mail/base/content/SearchDialog.js
--- a/mail/base/content/SearchDialog.js
+++ b/mail/base/content/SearchDialog.js
@@ -42,8 +42,11 @@ var gSearchStopButton;
 
 // Should we try to search online?
 var gSearchOnline = false;
 
+// Search all mail folders?
+var gSearchAllFolders;
+
 window.addEventListener("load", searchOnLoad);
 window.addEventListener("unload", () => {
   onSearchStop();
   searchOnUnload();
@@ -1106,8 +1109,17 @@ function searchOnLoad() {
   if (window.arguments && window.arguments[0]) {
     updateSearchFolderPicker(window.arguments[0].folder);
   }
 
+  if (
+    gCurrentFolder &&
+    (gCurrentFolder.server.type == "rss" ||
+      gCurrentFolder.server.type == "nntp")
+  ) {
+    // All accounts search is only available if we start with a mail folder.
+    gSearchAllFolders.hidden = true;
+  }
+
   // Trigger searchTerm.js to create the first criterion.
   onMore(null);
   // Make sure all the buttons are configured.
   UpdateMailSearch("onload");
@@ -1135,8 +1147,27 @@ function initializeSearchWindowWidgets()
 
   // functionality to enable/disable buttons using nsSearchResultsController
   // depending of whether items are selected in the search results thread pane.
   top.controllers.insertControllerAt(0, nsSearchResultsController);
+
+  gSearchAllFolders = document.getElementById("checkSearchAllFolders");
+  gSearchAllFolders.checked = false;
+  gSearchAllFolders.addEventListener("click", searchAllFoldersClick);
+}
+
+function searchAllFoldersClick() {
+  if (gSearchAllFolders.checked) {
+    document.getElementById("searchHeading").disabled = true;
+    document.getElementById("searchableFolders").disabled = true;
+    document.getElementById("checkSearchSubFolders").checked = true;
+    document.getElementById("checkSearchSubFolders").disabled = true;
+    document.getElementById("saveAsVFButton").disabled = true;
+  } else {
+    document.getElementById("searchHeading").disabled = false;
+    document.getElementById("searchableFolders").disabled = false;
+    document.getElementById("checkSearchSubFolders").disabled = false;
+    document.getElementById("saveAsVFButton").disabled = false;
+  }
 }
 
 /**
  * Handle click on the gSearchStopButton button (when that's labeled "Stop").
@@ -1253,8 +1284,18 @@ function getSearchTerms() {
  */
 function getSearchFolders() {
   const searchFolders = [];
 
+  if (gSearchAllFolders.checked) {
+    // Get all mail folders.
+    for (const folder of MailServices.accounts.allFolders) {
+      if (folder.server.type != "rss" && folder.server.type != "nntp") {
+        searchFolders.push(folder);
+      }
+    }
+    return searchFolders;
+  }
+
   if (!gCurrentFolder.isServer && !gCurrentFolder.noSelect) {
     searchFolders.push(gCurrentFolder);
   }
 
diff --git a/mail/base/content/SearchDialog.xhtml b/mail/base/content/SearchDialog.xhtml
--- a/mail/base/content/SearchDialog.xhtml
+++ b/mail/base/content/SearchDialog.xhtml
@@ -36,8 +36,9 @@
   <link rel="stylesheet" href="chrome://messenger/skin/themeableDialog.css" />
   <link rel="stylesheet" href="chrome://messenger/skin/colors.css" />
   <link rel="stylesheet" href="chrome://messenger/skin/folderMenus.css" />
   <link rel="stylesheet" href="chrome://messenger/skin/contextMenu.css" />
+  <link rel="localization" href="messenger/messenger.ftl" />
   <script defer="defer" src="chrome://messenger/content/globalOverlay.js"></script>
   <script defer="defer" src="chrome://global/content/editMenuOverlay.js"></script>
   <script defer="defer" src="chrome://messenger/content/searchWidgets.js"></script>
   <script defer="defer" src="chrome://messenger/content/mailWindow.js"></script>
@@ -86,9 +87,15 @@
 
     <vbox id="searchTerms" class="themeable-brighttext" persist="height">
       <vbox>
         <hbox align="center">
-          <label value="&searchHeading.label;" accesskey="&searchHeading.accesskey;"
+          <checkbox id="checkSearchAllFolders"
+                    data-l10n-id="search-all-mail-accounts"
+                    checked="false"
+                    persist="checked"/>
+        </hbox>
+        <hbox align="center">
+          <label id="searchHeading" value="&searchHeading.label;" accesskey="&searchHeading.accesskey;"
                  control="searchableFolders"/>
           <menulist id="searchableFolders" class="folderMenuItem"
                     displayformat="verbose">
             <menupopup is="folder-menupopup" class="menulist-menupopup"
diff --git a/mail/locales/en-US/messenger/messenger.ftl b/mail/locales/en-US/messenger/messenger.ftl
--- a/mail/locales/en-US/messenger/messenger.ftl
+++ b/mail/locales/en-US/messenger/messenger.ftl
@@ -589,4 +589,9 @@ quota-panel-percent-used = { $percent }%
 mark-as-read-action = Mark as Read
 delete-action = Delete
 mark-as-starred-action = Mark as Starred
 mark-as-spam-action = Mark as Spam
+
+## Hack for Betterbird: Search all mail accounts.
+search-all-mail-accounts =
+  .label = Search all mail accounts
+  .accesskey = m
